codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java

```diff
@@ -267,20 +277,67 @@ protected void handlerRemoved0(ChannelHandlerContext ctx) throws Exception { }
     public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
         if (msg instanceof ByteBuf) {
             CodecOutputList out = CodecOutputList.newInstance();
+            ByteBuf data = (ByteBuf) msg;
             try {
-                ByteBuf data = (ByteBuf) msg;
-                first = cumulation == null;
-                if (first) {
-                    cumulation = data;
-                } else {
+                do {
+                    first = cumulation == null;
+                    if (first) {
+                        // transfer data ownership to cumulation, so other methods have access to it
+                        cumulation = data;
+                        data = null;
+
+                        callDecode(ctx, cumulation, out);
+                        if (cumulation == null || !cumulation.isReadable()) {
```

Comment:
@belliottsmith I think you can remove the null check as it can never be null.

Reply:
I think it can - if the handler is removed before the callDecode completes, this will release the cumulation

Reply:
Thats a good point. Please add a comment and also add a null check here: https://github.com/netty/netty/pull/8890/files#diff-150d8a877292b8812bb4e3d27fb00a7fR314